tokio/src/scope/cancellation_token.rs

```diff
@@ -0,0 +1,1186 @@
+    }
+}
+
+/// The maximum permitted amount of references to a CancellationToken. This
+/// is derived from the intent to never use more than 32bit in the `Snapshot`.
+const MAX_REFS: u32 = (std::u32::MAX - 7) >> 3;
+
+/// Internal state of the `CancellationToken` pair above
+struct CancellationTokenState {
+    state: AtomicUsize,
+    parent: Option<NonNull<CancellationTokenState>>,
+    from_parent: SynchronizedThroughParent,
+    synchronized: Mutex<SynchronizedState>,
+}
+
+impl CancellationTokenState {
+    fn new(
+        parent: Option<NonNull<CancellationTokenState>>,
+        state: StateSnapshot,
+    ) -> CancellationTokenState {
+        CancellationTokenState {
+            parent,
+            from_parent: SynchronizedThroughParent {
+                prev_peer: None,
+                next_peer: None,
+            },
+            state: AtomicUsize::new(state.pack()),
+            synchronized: Mutex::new(SynchronizedState::new()),
+        }
+    }
+
+    /// Returns a snapshot of the current atomic state of the token
+    fn snapshot(&self) -> StateSnapshot {
+        StateSnapshot::unpack(self.state.load(Ordering::SeqCst))
+    }
+
+    fn atomic_update_state<F>(&self, mut current_state: StateSnapshot, func: F) -> StateSnapshot
+    where
+        F: Fn(StateSnapshot) -> StateSnapshot,
+    {
+        let mut current_packed_state = current_state.pack();
+        loop {
+            let next_state = func(current_state);
+            match self.state.compare_exchange(
+                current_packed_state,
+                next_state.pack(),
+                Ordering::SeqCst,
+                Ordering::SeqCst,
+            ) {
+                Ok(_) => {
+                    return next_state;
+                }
+                Err(actual) => {
+                    current_packed_state = actual;
+                    current_state = StateSnapshot::unpack(actual);
+                }
+            }
+        }
+    }
+
+    fn increment_refcount(&self, current_state: StateSnapshot) -> StateSnapshot {
+        self.atomic_update_state(current_state, |mut state: StateSnapshot| {
+            if state.refcount >= MAX_REFS as usize {
+                eprintln!("[ERROR] Maximum reference count for CancellationToken was exceeded");
+                std::process::abort();
+            }
+            state.refcount += 1;
+            state
+        })
+    }
+
+    fn decrement_refcount(&self, current_state: StateSnapshot) -> StateSnapshot {
+        let current_state = self.atomic_update_state(current_state, |mut state: StateSnapshot| {
+            state.refcount -= 1;
+            state
+        });
+
+        // Drop the State if it is not referenced anymore
+        if !current_state.has_refs() {
+            // Safety: `CancellationTokenState` is always stored in refcounted
+            // Boxes
+            let _ = unsafe { Box::from_raw(self as *const Self as *mut Self) };
+        }
+
+        current_state
+    }
```

Comment:
I'm a little unsure about the box and custom ref counting here. At a glance, it seems like we _should_ be able to just use `Arc` for this? The `CancellationToken` struct could own an `Arc<CancellationTokenState>`, and the `parent` field on `CancellationTokenState` would be an `Option<Arc<...>>`.

AFAICT, this would simplify the implementation a bit, and remove the need for some unsafe code here not directly related to the linked list. It _might_ also be slightly faster, because `Arc`s reference count increment and decrement operations are less strongly ordered (and don't need compare and swap loops since the reference count isn't packed with other state). Separating the ref count might also make our CAS loops less likely to fail, since refcount ops wouldn't invalidate the snapshot, so we might spin fewer times. 

Is there a reason that we can't use `Arc` that I'm overlooking? If so, it would be nice for that to be documented.

Reply:
We need a weak pointer like mechanism, since the parent references the childs (in order to be able to cancel them when the parent cancels) and the child owns a parent pointer (e.g. for sibling lists). With weak pointers we would leak everything. 

It's unfortunately a rather hard problem. That was why the original tests were failing.

Reply:
Ah, okay. I'll have to take a look at earlier versions of this at some point, thanks.